.gitignore

```diff
@@ -430,6 +430,11 @@ miktex*.zip
 **/lgb.Dataset.data
 **/model.txt
 **/lgb-model.txt
+tests/distributed/mlist.txt
```

Comment:
We need to replace all `open(smth)` calls with
```
CURR_DIR = os.path.abspath(os.path.dirname(__file__))

...

open(os.path.join(CURR_DIR, smth))
```
to be sure that these lines in `.gitignore` are correct. Because now all files are created in the directory from which `pytest` is called.

Reply:
I had thought about this as well. Can we define if `pytest` is going to be called from the top level directory (`BUILD_DIRECTORY` in [test.sh](https://github.com/microsoft/LightGBM/blob/master/.ci/test.sh)) or from some other path? Right now everything happens in `tests/distributed`, I even had to copy the binary there. I think it may make more sense to call `pytest` from the top level directory and modify the paths here so that `pytest` is called where the binary is and the files get created at `tests/distributed`. Is that what you suggested?

Reply:
I added this idea in f7c6fddc2e82ea55773e72f7639ff5d6514a0cb4, please let me know what you think. `BINARY_DIR` would be `BUILD_DIRECTORY` and `TESTS_DIR` would be `BUILD_DIRECTORY/tests/distributed`

Reply:
I suppose we can add an argument with `lightgbm[.exe]` executable file location and don't force that file to be in `distributed` directory.

https://stackoverflow.com/a/55301318

Reply:
In f7c6fdd the binary is assumed to be where it gets built (`BUILD_DIRECTORY`), so there's no need to move it anymore and the tests can be run with `pytest tests/distributed/_test_distributed.py`

Reply:
Executable file is not always located in the `BUILD_DIRECTORY` folder. On Windows, for example, it is in `BUILD_DIRECTORY/windows/x64/Release` folder or `BUILD_DIRECTORY/Release` folder depending on how (via command line or Visual Studio GUI) executable was built.

Reply:
I thought this was going to be tested on linux only. Will look into passing the executable location to pytest

Reply:
Since this PR can be treated as some kind of framework for testing distributed training, I believe more general solutions here are important.

> Will work in passing the executable location to pytest

Thank you very much!

Reply:
Added in 56113de, the argument is `--execdir`, so the tests could be run with `pytest tests/distributed/_test_distributed.py --execdir BUILD_DIRECTORY/Release`. Looking forward to your thoughts. I haven't used lightgbm on windows so I'm not sure if the executable has the `.exe` suffix

Reply:
> I'm not sure if the executable has the `.exe` suffix

You're right - it has. But suffix can be omitted in calls. Both lines below result in opening Edge browser:

![image](https://user-images.githubusercontent.com/25141164/121266686-a48fd680-c8c3-11eb-8a9a-0c8631ad0bb4.png)


Just a question: why did you decide to introduce `execdir` argument and deduce full path to the executable later in tests but not directly `execfile`?

Reply:
Oh I didn't think of that haha, I'll change it to execfile.